[Key Vault] 4.1.0 CHANGELOG.md and version updates (#10477)
Azure/azure-sdk-for-js: sdk/keyvault/keyvault-certificates/CHANGELOG.md

# Release History

## 4.1.0 (Unreleased)

> - Added the optional `serviceVersion` property to the `CertificateClient` optional parameters.
>   It defaults to the latest supported API version, which currently is `7.1`.

Reviewer: ```suggestion
- Added the optional `serviceVersion` property to the `CertificateClient` optional parameters.
      - It defaults to the latest service version, which currently is `7.1`.
      - Other supported service version at the moment is `7.0`.
      - If you were using `4.1.0-preview.1` version of this package, then this was previously named `apiVersion`.
```

Same feedback for the other 2 packages

Author: Thank you. I thought you didn't want the preview package to be mentioned, but I get it now! I like this.

> - Added `recoverableDays` as an optional property to `KeyProperties`.

Reviewer: ```suggestion
- Added `recoverableDays` as an optional property to `CertificateProperties`.
```

> - Fixed a bug on `importCertificate`. The bytes of a PEM formatted certificate were being encoded again on base64.

Reviewer: Is there an issue or PR we can link that would have more details?

  Now, if the user provides a policy with contentType `application/x-pem-file`, the bytes will be assumed to be formatted using ASCII,
  and thus will be sent as is.
- Now using `Poller` and `PollerLike` from the latest `@azure/core-lro` instead of `KVPoller` and `KVPollerLike`.
- `KVPollerLike` is now an alias of `PollerLike`.

> - `KVPollerLike` is considered deprecated. Use `PollerLike`.

Reviewer: Can we make the above 2 lines sub bullet points for the first line on the poller?

> 

Reviewer: https://gist.github.com/sadasant/2cb528aa8502e1b14fd1552516bacafb/revisions shows a `CertificateKeyType` supporting a new type `oct`. Should we talk about that here?

Author: Yes! Thank you.

Author: I've updated this, please let me know how it looks!

# Release History

## 4.1.0 (2020-08-11)

> - Added the optional `serviceVersion` property to the `CertificateClient` optional parameters.

Reviewer: ```suggestion
- Added the optional `serviceVersion` property to the `CertificateClient` optional parameters to control the version of the Key Vault service being used by the client. 
```

# Release History

## 4.1.0 (2020-08-11)

- Added the optional `serviceVersion` property to the `CertificateClient` optional parameters.
    - It defaults to the latest supported API version, which currently is `7.1`.
    - Other supported service version at the moment is `7.0`.
    - If you were using `4.1.0-preview.1` version of this package, then this was previously named `apiVersion`.
- `CertificateKeyType` now supports the `oct` type. To align with the Key Vault Keys client.
- Added `recoverableDays` as an optional property to `CertificateProperties`.

>     - When creating and updating a certificate, set the `recoverableDays` property to define the number of days in which the certificate can be recovered after deletion. When certificates are read from the service, they will have the `recoverableDays` property in their `properties` field.

Reviewer: > When certificates are read from the service, they will have the `recoverableDays` property in their `properties` field

I am not sure if "read from the service" makes it clear that we are talking about the `get` methods. I think we can skip it altogether assuming that anything that is set when creating/updating can be read when using the `get` methods.

```suggestion
    - When creating or updating a certificate, set the `recoverableDays` property to define the number of days in which the certificate can be recovered after deletion.
```

Reviewer: Same for keys & secrets

Author: I decided to not go with this suggestion because this `create or update` section is invalid. This property is read-only. In the process of analyzing this, I made and closed this PR: https://github.com/Azure/azure-sdk-for-js/pull/10543 even though recoverableDays is part of the properties now, it's read only for all of the clients. Keys and Secrets don't extend the properties when creating entities (keys or secrets), only Certificates does this (but the service won't be reading it if it's sent on the `beginCreateCertificate` method either).

# Release History

## 4.1.0 (2020-08-11)

- Added the optional `serviceVersion` property to the `CertificateClient` optional parameters.
    - It defaults to the latest supported API version, which currently is `7.1`.
    - Other supported service version at the moment is `7.0`.
    - If you were using `4.1.0-preview.1` version of this package, then this was previously named `apiVersion`.
- `CertificateKeyType` now supports the `oct` type.

>     - Since the service now supports `oct` keys, now the client will correctly report the type of these keys.  

Reviewer: Is there any service documentation we can link to for getting more details on what this `oct` key is? I saw somewhere this being referred to as `Octet`. What is the right term to use?

- Added `recoverableDays` as an optional property to `CertificateProperties`.

>     - The Key Vault Keys client `4.1.0` now allows for Keys to be created with a determined number of days in which they can be recovered after deletion.
>       For this reason, the Certificates client will now show this `recoverableDays` property when reading the key in use. This is only relevant for Azure Key Vaults with the soft-delete setting enabled.

Reviewer: We are talking about Keys client here, it should be certificates.
We don't necessarily need to call out `4.1.0` here in the sub bullet point, as that is what the whole section is about.
I would recommend the below

```suggestion
    - Use this when creating a certificate to provide the number of days in which a certificate can be recovered after deletion. The same property will reflect the set value when reading the certificates.
    - This is applicable only for Azure Key Vaults with the soft-delete setting enabled.
```

Author: I used this instead:

```
    - When creating and updating a certificate, use the `recoverableDays` property to set the number of days in which the certificate can be recovered after deletion. When certificates are read from the service, they will have the `recoverableDays` property in their `properties` field.
    - This is applicable only for Azure Key Vaults with the soft-delete setting enabled.
```

thoughts?

- Now using `Poller` and `PollerLike` from the latest `@azure/core-lro` instead of `KVPoller` and `KVPollerLike`.
    - `KVPollerLike` is now an alias of `PollerLike`.
    - `KVPollerLike` is considered deprecated. Use `PollerLike`.

> - Fixed [bug 7407](https://github.com/Azure/azure-sdk-for-js/issues/7407) on `importCertificate`.
>   The bytes of a PEM formatted certificate were being encoded again on base64.
>   Now, if the user provides a policy with contentType `application/x-pem-file`, the bytes will be assumed to be formatted using ASCII,
>   and thus will be sent as is.

Reviewer: We don't really send it "as is" right? We convert the bytes to string and then send it

```suggestion
- If the policy in the options has contentType `application/x-pem-file` when using the `importCertificate`
  method, we now encode the bytes of the certificate as an ASCII string instead of base64 which is the default
  treatment. This is to fix [bug 7407](https://github.com/Azure/azure-sdk-for-js/issues/7407)
```

# Release History

> ## 4.1.0-preview.2 (Unreleased)

Reviewer: We should not be changing the changelog that was put in place for past releases.
Add a new section for the GA version and note the net changes compared to the previous GA version

Reviewer: My mistake. I didnt realize that 4.1.0-preview.2 was not released. Therefore, it makes perfect sense that we change this header to from `4.1.0-preview.2` to `4.1.0`

> ## 4.1.0 (Unreleased)

Reviewer 2: I've heard it recommended you don't remove the preview version until you're ready to release, in case it accidentally leaks out (burns a version number).

Author: I have copied to 4.1.0 only the changes that have happened after the last non-preview version. I have also restored the existing information, but I realized that I forgot to add to master the changelog entries for the last non-preview versions, so I have added those.

Reviewer: > I've heard it recommended you don't remove the preview version until you're ready to release, in case it accidentally leaks out (burns a version number).

This is us getting ready to release though :)
Making all the changes so that we can release on Monday/Tuesday

Author: Ok I removed the unreleased preview sections! Thank you 🌞

- Added the optional `serviceVersion` property to the `CertificateClient` optional parameters.
  It defaults to the latest supported API version, which currently is `7.1`.
- Added `recoverableDays` as an optional property to `KeyProperties`.
- Fixed a bug on `importCertificate`. The bytes of a PEM formatted certificate were being encoded again on base64.
  Now, if the user provides a policy with contentType `application/x-pem-file`, the bytes will be assumed to be formatted using ASCII,

>   and thus will be sent as is.

Reviewer: The 3 bug fixes listed below were part of 4.0.2 release for certificates. So, they should not be appearing again for 4.1.0

Author: Thank you for your feedback! 🌞

I have copied to 4.1.0 only the changes that have happened after the last non-preview version. I have also restored the existing information, but I realized that I forgot to add to master the changelog entries for the last non-preview versions, so I have added those.

Feedback appreciated!

# Release History

## 4.1.0 (Unreleased)

> - Added support for service version 7.1.

Reviewer: Perhaps there is some value talking about the fact that we do support both versions? Or has this been covered in previous updates?

```suggestion
- Added support for service version 7.1 which will now be the default version used by this package. You can target
the older 7.0 version using the `apiVersion` in the options passed to the client constructor.
```

Author: Sounds good, I'll think of an alternative phrasing and I'll send it to you to make it faster.

Author: I'm resolving this, since these files have changed considerably.

    - It defaults to the latest supported API version, which currently is `7.1`.
    - Other supported service version at the moment is `7.0`.
    - If you were using `4.1.0-preview.1` version of this package, then this was previously named `apiVersion`.
- `CertificateKeyType` now supports the `oct` type. To align with the Key Vault Keys client.
- Added `recoverableDays` as an optional property to `CertificateProperties`.

>     - When creating and updating a certificate, use the `recoverableDays` property to set the number of days in which the certificate can be recovered after deletion. When certificates are read from the service, they will have the `recoverableDays` property in their `properties` field.

Reviewer 2: Nit: "use the recoverableDays property to set the" -> "set the recoverableDays property"
